### PR TITLE
DigitalTwins: Switch to etags, drop tiny Options classes

### PR DESCRIPTION
**Pull request by tg-msft**

### CODE REVIEW

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwin.cs [hunk 1]
```diff
@@ -86,7 +88,26 @@ public class BasicDigitalTwin
         /// A string representing a weak ETag for the entity that this request performs an operation against, as per RFC7232.
         /// </summary>
         [JsonPropertyName(DigitalTwinsJsonPropertyNames.DigitalTwinETag)]
-        public string ETag { get; set; }
+        [JsonConverter(typeof(OptionalETagConverter))] // TODO: Remove when #16272 is fixed
```
Comment (drwill-ms) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwin.cs:91]: this is interesting

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/PublishTelemetryTests.cs [hunk 2]
```diff
@@ -47,26 +47,18 @@ public async Task PublishTelemetry_Lifecycle()
                 await CreateModelsAndTwins(client, wifiModelId, roomWithWifiModelId, wifiComponentName, roomWithWifiTwinId).ConfigureAwait(false);
 
                 // Act - Test publishing telemetry to a digital twin.
-                var telemetryOptions = new PublishTelemetryOptions()
-                {
-                    TimeStamp = default
-                };
-                Response publishTelemetryResponse = await client.PublishTelemetryAsync(roomWithWifiTwinId, Recording.Random.NewGuid().ToString(), "{\"Telemetry1\": 5}", telemetryOptions).ConfigureAwait(false);
+                Response publishTelemetryResponse = await client.PublishTelemetryAsync(roomWithWifiTwinId, Recording.Random.NewGuid().ToString(), "{\"Telemetry1\": 5}", timeStamp: default(DateTimeOffset)).ConfigureAwait(false);
```
Comment (azabbasi) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/tests/PublishTelemetryTests.cs:50]: nit: line seems too long. we can break it into multiple

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/PublishTelemetryTests.cs [hunk 3]
```diff
@@ -47,26 +47,18 @@ public async Task PublishTelemetry_Lifecycle()
...
-                {
-                    TimeStamp = default
-                };
                 var telemetryPayload = new Dictionary<string, int>
                 {
                     { "ComponentTelemetry1", 9}
                 };
                 Response publishComponentTelemetryResponse = await client
-                    .PublishComponentTelemetryAsync(roomWithWifiTwinId, wifiComponentName, Recording.Random.NewGuid().ToString(), JsonSerializer.Serialize(telemetryPayload), componentTelemetryOptions)
+                    .PublishComponentTelemetryAsync(roomWithWifiTwinId, wifiComponentName, Recording.Random.NewGuid().ToString(), JsonSerializer.Serialize(telemetryPayload), timeStamp: default(DateTimeOffset))
```
Comment (azabbasi) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/tests/PublishTelemetryTests.cs:61]: same here.

**Comment by tg-msft** [comment]
It's also worth pointing out that we can switch from ["<= 6 parameters" simple params to an Options type complex param](https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-parameters) in the future without breaking if the service adds many more parameters.

Assuming we had something today like:
```C#
public void Foo(string id, bool include = false) { ... }
```

We can switch to an Options bag in the future with:
```C#
[EditorBrowsable(Never)]
public void Foo(string id, bool include) { ... } // note: default value removed

public void Foo(string id, FooOptions options = default) { ... }

public class FooOptions
{
    public bool Include { get; set; }
    public string Random { get; set; }
    // + more new params...
}
```

And it would not be a breaking change.  Customers will continue to compile, upgrading without recompile is fine, etc.  We'd only see the new signature in intellisense and documentation so new customers would always start there.  The only pain is customers wanting to use the new `Random` property would have to switch from `Foo(id, include: true)` to `Foo(id, new FooOptions { Include = true, Random = "..." })`.  We consider that acceptable and have done it often in other services like Storage.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/PublishTelemetryTests.cs (hunk 2 above)
Comment (tg-msft) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/tests/PublishTelemetryTests.cs]: Split across lines

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/PublishTelemetryTests.cs (hunk 3 above)
Comment (tg-msft) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/tests/PublishTelemetryTests.cs]: Split across lines

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwin.cs (hunk 1 above)
Comment (tg-msft) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwin.cs:91]: (And let's reference #16272 here so it shows up on the issue too)

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/DigitalTwinsClient.cs [hunk 4]
```diff
@@ -1931,7 +2017,7 @@ public virtual Response DeleteEventRoute(string eventRouteId, CancellationToken
             string digitalTwinId,
             string messageId,
             string payload,
-            PublishTelemetryOptions options = null,
+            DateTimeOffset? timeStamp = null,
```
Comment (drwill-ms) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/src/DigitalTwinsClient.cs:2020]: I wonder if this should be cased `timestamp`. The only other case of timestamp I found in .NET is `Stopwatch.GetTimestamp`.

https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.stopwatch.gettimestamp?view=netcore-3.1

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/DigitalTwinsClient.cs [hunk 5]
```diff
@@ -1940,6 +2026,12 @@ public virtual Response DeleteEventRoute(string eventRouteId, CancellationToken
                 messageId = Guid.NewGuid().ToString();
             }
 
+            PublishTelemetryOptions options = new PublishTelemetryOptions();
```
Comment (drwill-ms) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/src/DigitalTwinsClient.cs:2029]: var

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwin.cs [hunk 6]
```diff
@@ -86,7 +88,26 @@ public class BasicDigitalTwin
...
+        public ETag? ETag { get; set; }
+
+        // TODO: Remove when #16272 is fixed
+        private class OptionalETagConverter : JsonConverter<ETag?>
+        {
+            public override ETag? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+            {
+                string value = reader.GetString();
+                return value != null ?
```
Comment (drwill-ms) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwin.cs:100]:
```csharp
return value != null
    ? new ETag(value)
    : (ETag?)null;
```

**Review by drwill-ms — APPROVED** [review]

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/DigitalTwinsClient.cs (hunk 4 above)
Comment (tg-msft) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/src/DigitalTwinsClient.cs]: Krzysztof wrote `Stopwatch`.  I'll probably get yelled at if we don't follow it.  :smile:

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/DigitalTwinsClient.cs (hunk 5 above)
Comment (tg-msft) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/src/DigitalTwinsClient.cs]: Fixed.

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwin.cs (hunk 6 above)
Comment (tg-msft) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicDigitalTwin.cs]: Fixed.

**Review by barustum — APPROVED** [review]

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/Serialization/OptionalETagConverter.cs [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+            string value = reader.GetString();
+            return value != null
+                ? new ETag(value)
+                : (ETag?)null;
+        }
+
+        public override void Write(Utf8JsonWriter writer, ETag? value, JsonSerializerOptions options)
+        {
+            if (value == null)
+            { writer.WriteNullValue(); }
```
Comment (drwill-ms) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/src/Serialization/OptionalETagConverter.cs:26]: won't this NRE?

File: sdk/digitaltwins/Azure.DigitalTwins.Core/tests/ComponentTests.cs [hunk 8]
```diff
@@ -93,5 +93,191 @@ public async Task Component_Lifecycle()
...
+                        componentUpdatePatchDocument)
+                    .ConfigureAwait(false);
+
+                // Get the latest ETag
+                ETag? etagBeforeUpdate = (await client.GetDigitalTwinAsync<BasicDigitalTwin>(roomWithWifiTwinId)).Value.ETag;
+                Assert.IsNotNull(etagBeforeUpdate);
+
+                // Patch component again, but with the now out of date ETag
+                JsonPatchDocument secondComponentUpdatePatchDocument = new JsonPatchDocument();
+                componentUpdatePatchDocument.AppendReplace("/Network", "Even newer Network");
```
Comment (drwill-ms) [inline sdk/digitaltwins/Azure.DigitalTwins.Core/tests/ComponentTests.cs:241]: Setting the wrong JsonPatchDocument here

File: sdk/digitaltwins/Azure.DigitalTwins.Core/src/Serialization/OptionalETagConverter.cs (hunk 7 above)
Comment (tg-msft) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/src/Serialization/OptionalETagConverter.cs:26]: If the `writer` is null?  That won't happen with STJ and it's not a publicly exposed type.  We can add an `Argument.NotNull(writer, nameof(writer));` if we think it's possible customers could somehow get here.

Comment (drwill-ms) [reply sdk/digitaltwins/Azure.DigitalTwins.Core/src/Serialization/OptionalETagConverter.cs:26]: Ah, I misread this as `if (write == null)` for some reason.

**Pull request merged**